samza-core/src/main/scala/org/apache/samza/checkpoint/OffsetManager.scala

```diff
@@ -231,24 +231,64 @@ class OffsetManager(
-        sspToOffsets.asScala.filterKeys(sspsForTaskName.contains)
-      } else {
-        warn(taskName + " is not found... ")
-        Map[SystemStreamPartition, String]()
-      }
+      val taskStartingOffsets = startingOffsets.getOrElse(taskName,
+        throw new SamzaException("Couldn't find starting offsets for task: " + taskName))
+
+      val taskSSPs = systemStreamPartitions.getOrElse(taskName,
+        throw new SamzaException("No SSPs registered for task: " + taskName))
 
-      new Checkpoint(new HashMap(partitionOffsets.asJava)) // Copy into new Map to prevent mutation
+      // filter the offsets in case the task model changed since the last checkpoint was written.
+      val taskLastProcessedOffsets =
+        lastProcessedOffsets.getOrDefault(taskName, new ConcurrentHashMap()).asScala
+          .filterKeys(taskSSPs.contains)
+
+      val modifiedTaskOffsets = getModifiedOffsets(taskStartingOffsets, taskLastProcessedOffsets)
+      new Checkpoint(new HashMap(modifiedTaskOffsets)) // Copy into new Map to prevent mutation
     } else {
       debug("Returning null checkpoint for taskName %s because no checkpoint manager/callback is defined." format taskName)
       null
     }
   }
 
+  /**
+    * Call the beforeCheckpoint() method on [[CheckpointListener]] SystemConsumers to give them a chance to inspect and
+    * potentially modify the offsets about to be checkpointed.
+    */
+  def getModifiedOffsets(taskStartingOffsets: Map[SystemStreamPartition, String],
+      taskLastProcessedOffsets: Map[SystemStreamPartition, String]): HashMap[SystemStreamPartition, String] = {
+    val modifiedOffsets = new HashMap[SystemStreamPartition, String]()
+    modifiedOffsets.putAll(taskLastProcessedOffsets.asJava)
+
+    taskLastProcessedOffsets
+      .groupBy { case (ssp, offset) => ssp.getSystem }  // Map(systemName -> Map(ssp -> offset))
+      .foreach { case (systemName, systemSSPLastProcessedOffsets) =>
+      if (checkpointListeners.contains(systemName)) {
+        val checkpointListener = checkpointListeners(systemName)
+
+        // if we haven't processed any messages from the system since we started, we don't need to call
+        // the checkpoint listener. also, in case of Kafka, getting the safe offset is only possible after
+        // first poll() is successful. check if we need to adjust the offset for any of the ssp in the system.
+        val needModifiedOffsets = systemSSPLastProcessedOffsets.exists { case (ssp, sspLastProcessedOffset) =>
+          // starting offsets are checkpointed offsets + 1 (see 'loadStartingOffsets'),
+          // while on container start last processed offsets are equal to checkpointed offsets.
+          // subtract one from starting offsets to compare them with last processed offsets.
+          val adjustedSSPStartingOffset = taskStartingOffsets.get(ssp) match {
+            case Some(offset) if offset != null => (offset.toLong - 1).toString
```

Comment:
There's a hidden assumption here on contiguous offsets - since, we're by-passing the offset comparator. maybe that's ok?

Reply:
That's true, however there's no SystemAdmin#getOffsetsBefore() API that we can use here.

This is ported from existing samza-li code for using li-apache-kafka-consumer with safe offsets. I think this was a workaround for the KafkaConsumer specific behavior that poll() must be called on the consumer before requesting a safe offset.

Now that you mention it, I think the bigger issue is that this is also assuming that the offset is a long, which is definitely not true in general. I can check if we can handle the KafkaConsumer specific issue by catching whatever exception safeOffset returns.